mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala

```diff
@@ -1110,4 +1124,40 @@ private[spark] object RandomForest extends Logging {
     }
   }
 
+  /**
+   * Class for queueing nodes to split on each iteration.
+   * This is a FILO queue.
+   */
+  private[impl] class NodeQueue {
+    private var q: mutable.MutableList[(Int, LearningNode)] =
+      mutable.MutableList.empty[(Int, LearningNode)]
+
+    /** Put (treeIndex, node) into queue. Constant time. */
+    def put(treeIndex: Int, node: LearningNode): Unit = {
+      q += ((treeIndex, node))
+    }
+
+    /** Remove and return last inserted element.  Linear time (unclear in Scala docs though) */
```

Comment:
I also somehow didn't find Stack via Google...  Does anyone know of documentation for the performance of Stack?  I don't see it in the Scala docs.